### PR TITLE
Pull project-related changes from fst3a/feature/impl-file

### PR DESCRIPTION
**Pull request by nanokatze**

### CODE REVIEW

File: spade/core/project.py [hunk 1]
```diff
@@ -111,3 +77,65 @@ def get_template_for_file(self, f, template):
...
+        conn = self.__db_engine.connect()
+        conn.execute(ins, key=key, value=value) # TODO: might cause problems?
+
+    def __init_db(self, engine):
+        """
+        Creates default tables for a newly created spade project
+        """
+        file_hash_len = 256/8
+        change_hash_len = 256/8
+        metadata = MetaData()
```
Comment (nyxxxie) [inline spade/core/project.py]: When are these vars ever used?

File: spade/core/project.py [hunk 2]
```diff
@@ -1,55 +1,17 @@
...
+from sqlalchemy import create_engine, MetaData, Table, Column, Binary, Integer, String, ForeignKey
 
-        # Define files table
-        self.table_files = Table("files", metadata,
-            Column("id", Integer, primary_key=True, autoincrement=True),
-            Column("path", String, unique=True),                # Relative or absolute path to this file
-            Column("hash", Binary),                             # sha256 hash of file contents on last change
-            Column("head_change", Binary, server_default=None), # head change file is at
-        )
+SCHEMA_VERSION = "0.0a"
```
Comment (nyxxxie) [inline spade/core/project.py]: You define this var but then never use it (it should be used in constructor where we set schema version project info variable).  Also I don't want to use a strict versioning system with major and minor changes for this, lets just use a nonce and up the number each time we change the schema.

File: spade/core/project.py [hunk 3]
```diff
@@ -59,6 +21,9 @@ def __init__(self, dbfile):
         self.__add_info("creation_datetime", datetime.datetime.now(), True)
         self.__add_info("update_datetime", datetime.datetime.now())
 
+    def db(self):
```
Comment (nyxxxie) [inline spade/core/project.py]: We never define a _db object, I instead have a function db_engine() that returns the sqlalchemy engine.

File: spade/core/project.py [hunk 4]
```diff
@@ -111,3 +77,65 @@ def get_template_for_file(self, f, template):
...
+            Column("parent", Binary),
+            # position in file where change occured
+            Column("file_pos", Integer),
+            # NOTE(fst3a): Already solved this
+            # change type.  '+' = insert, '-' = erase, '!' = replace
+            Column("change_type", Integer),
+            # bytes that were inserted or erased
+            Column("change", Binary)
+        )
+        # FIXME(fst3a): This is plain wrong, one project can have multiple files
```
Comment (nyxxxie) [inline spade/core/project.py]: If you're undecided on how this feature should function, perhaps we should just make it a seperate feature and merge it when it is ready?  I don't like large amounts of commented out code accompanied by a large FIXME comment just sitting in code that might be merged soon.

**Review by nyxxxie — CHANGES_REQUESTED** [review]
Left comments, check them out and comment or address them

File: spade/core/project.py (hunk 1 above)
Comment (nanokatze) [reply spade/core/project.py]: whops, they aren't, forgot to remove

File: spade/core/project.py (hunk 2 above)
Comment (nanokatze) [reply spade/core/project.py]: It is a constant, not a variable.  It should be inserted in project_info on new db or upgrade.

File: spade/core/project.py (hunk 3 above)
Comment (nanokatze) [reply spade/core/project.py]: Ok.

File: spade/core/project.py (hunk 4 above)
Comment (nanokatze) [reply spade/core/project.py]: Ok.

File: spade/core/project.py [hunk 5]
```diff
@@ -14,10 +14,10 @@ class Project:
...
-        self.__dbfile = dbfile
-        self.__db_engine = create_engine("sqlite:///"+dbfile)
-        self.__init_db(self.__db_engine)
-        self.__add_info("schema_version", "1")
+        self._dbfile = dbfile
+        self._db_engine = create_engine("sqlite:///" + dbfile)
+        self._init_db(self._db_engine)
+        self._add_info("schema_version", SCHEMA_VERSION)
         date = datetime.datetime.now()
         self.__add_info("creation_datetime", date, True)
```
Comment (nyxxxie) [inline spade/core/project.py]: Missed a spot :^)

Comment (nanokatze) [reply spade/core/project.py]: *censored*

**Review by nyxxxie — APPROVED** [review]
Looks good

**Pull request merged**